### PR TITLE
make fancy filter work on user table

### LINKED ISSUE #3085: Use user filter object to filter Users table (Fancy Filter)
## Description

When the filter modal is closed, use the filter object to refresh the query.
Filter can be stored in state on the user table.

## Acceptance Criteria
- [ ] When the modal is opened, it is prepopulated with the current state.
- [ ] When the modal is closed, the filter state is saved and the table query refreshed.
- [ ] Cypress tests

### PR DESCRIPTION
**Pull request by vd1992**
closes #3085 
This makes the filter actually operate and changes the query. 
To make this work, I added a state in the parent component UserTable that is changed in the (grand) child component UserTableFilterDialog. That way, the filter change is received in UserTable enabling query re-run. 
I also combined the search bar filter with the fancy filter enabling both to work together. 
A single test was added the existing Jest file, not sure what useful testing could be added because not sure a Cypress test may help much as it is important to ensure the generated request matches the filtering selected by manual testing

Testing: 
go to /admin/users
pop open the network tab on the side
play around with the filtering, try filtering something then ensure it is mirrored accurately in a network request 
manual exercise filtering cases, the goal is to ensure the appropriate request is generated

### CODE REVIEW

File: frontend/admin/src/js/components/user/util.ts
Comment (patcon) [inline frontend/admin/src/js/components/user/util.ts:20]: This should be the same without explicitly mentioning each:

```suggestion
  if (Object.values(LanguageAbility).includes(selection as LanguageAbility)) {
    return selection as LanguageAbility;
  }
```

File: frontend/admin/src/js/components/user/util.ts
Comment (patcon) [inline frontend/admin/src/js/components/user/util.ts:50]:
```suggestion
  if (Object.values(WorkRegion).includes(selection as WorkRegion)) {
    return selection as WorkRegion;
  }
```

File: frontend/admin/src/js/components/user/util.ts
Comment (patcon) [inline frontend/admin/src/js/components/user/util.ts:86]:
```suggestion
  if (Object.values(OperationalRequirement).includes(selection as OperationalRequirement)) {
    return selection as OperationalRequirement;
  }
```

File: frontend/admin/src/js/components/user/util.ts
Comment (patcon) [inline frontend/admin/src/js/components/user/util.ts:101]:
```suggestion
  if (Object.values(JobLookingStatus).includes(selection as JobLookingStatus)) {
    return selection as JobLookingStatus;
  }
```

File: frontend/admin/src/js/components/user/util.ts
Comment (patcon) [reply frontend/admin/src/js/components/user/util.ts:20]: There maybe a more general way to do this in typescript (so one function works for all of these), but I'm unsure what this is off the top of my head :)

File: frontend/admin/src/js/components/user/UserTableFilterDialog.tsx
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:267]: Do you think this component needs 2 submission handlers? As in could this be moved into the onSubmit function that was already present? Maybe it was just poorly named, but `onSubmit` was the place where I assumed this logic would be injected/added

maybe `onSubmitFilter`? which identifies itself as "for a form" but also "for the filter" like you capture?

Comment (patcon) [reply frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:267]: EDIT: Wait. Nevermind, might have mixed myself up here. gimme a sec :)

**Comment by vd1992** [comment]
handy, merged it in, neat stuff
thanks @patcon

File: frontend/admin/src/js/components/user/UserTableFilterDialog.tsx
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:270]: I think we could prob adjust this type to only be string, which would let you clean this up a bit. We have some generic type variables of `Object<V>` but it's always string
```
# shows only string
git grep 'Option<[a-z]\+>'
```

For no particular reason I dragged forward an Option type from earlier that allowed key to be either string or number:
https://github.com/GCTC-NTGC/gc-digital-talent/blob/71ded63044636b25c0b5b0ccaad7a106eec85e34/frontend/common/src/components/form/Select/SelectFieldV2.tsx#L11

But I don't think we ever use a number, except in the stories for SelectField:

https://github.com/GCTC-NTGC/gc-digital-talent/blob/71ded63044636b25c0b5b0ccaad7a106eec85e34/frontend/common/src/components/form/Select/SelectFieldV2.stories.tsx#L38-L40

https://github.com/GCTC-NTGC/gc-digital-talent/blob/71ded63044636b25c0b5b0ccaad7a106eec85e34/frontend/common/src/components/form/MultiSelect/MultiSelectFieldV2.stories.tsx#L38-L40

File: frontend/admin/src/js/components/user/UserTableFilterDialog.tsx
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:287]: E.g.,

```suggestion
        skills: data.skills.map((skill) => ({ id: skill }),
```

File: frontend/admin/src/js/components/user/UserTable.tsx
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTable.tsx:87]: re: starting with verb, maybe something descriptive like

```suggestion
      where: buildUserFilterInput(fancyFilter, searchState?.term, searchState?.col),
```

File: frontend/admin/src/js/components/user/UserTable.tsx
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTable.tsx:90]: "column" feels like a leaky abstraction from the table

```suggestion
    searchType: string | undefined,
```

File: frontend/admin/src/js/components/user/UserTableFilterDialog.tsx
Comment (vd1992) [reply frontend/admin/src/js/components/user/UserTableFilterDialog.tsx]: yeah, I didn't want to be too destructive when it came to adding this pull request in so I tried to just adjust things to fit
typing changes https://github.com/GCTC-NTGC/gc-digital-talent/pull/3893/commits/70f4cfb7afae25937fa4320f5618497948611318

File: frontend/admin/src/js/components/user/UserTable.tsx
Comment (vd1992) [reply frontend/admin/src/js/components/user/UserTable.tsx]: sounds reasonable 
https://github.com/GCTC-NTGC/gc-digital-talent/pull/3893/commits/a2cb77df9d066a66a0701d8289082a4d0f68742f

File: frontend/admin/src/js/components/user/UserTable.tsx
Comment (vd1992) [reply frontend/admin/src/js/components/user/UserTable.tsx]: https://github.com/GCTC-NTGC/gc-digital-talent/commit/a2cb77df9d066a66a0701d8289082a4d0f68742f

File: frontend/admin/src/js/components/user/UserTableFilterDialog.tsx
Comment (vd1992) [reply frontend/admin/src/js/components/user/UserTableFilterDialog.tsx]: https://github.com/GCTC-NTGC/gc-digital-talent/commit/70f4cfb7afae25937fa4320f5618497948611318

File: frontend/admin/src/js/components/user/UserTableFilterDialog.tsx
Comment (patcon) [reply frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:268]: Bah, I'm fumbling to think how to merge these nicely, but do you see any way that a single function is possible instead of two? I think ideally the processing of formvalues into input type would happen in a single place, but I'm not seeing a clean way to do that. I recall onSubmit is a handler set up one level that doesn't do much, so feels odd to have two

Comment (vd1992) [reply frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:268]: if it is too much of a mess to attempt it, then it might not be worth the hassle frankly

Comment (patcon) [reply frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:268]: i'm onboard with that. extracting the processing out of the deep component is the hard part...

but what about just stripping out current `onSubmit` and renaming `onFilterChange` to `onSubmitFilter` (to hint at form submission and show up when grepping for "onSubmit")

Comment (patcon) [reply frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:268]: Oh! Wait... you could also use it to pass your function, and keep the processing out of the component:

```tsx
  const handleFilterSubmit: SubmitHandler<FormValues> = (data) => {
    onFilterChange( //...
    // ...
  };
```

**Comment by vd1992** [comment]
alright made some changes @patcon 
removed the function here https://github.com/GCTC-NTGC/gc-digital-talent/pull/3893/commits/af0b74fac45f4a1e5ece1323677786aec4824835 
and changed filter format to add education here https://github.com/GCTC-NTGC/gc-digital-talent/pull/3893/commits/b87e3543dae9a148d9b6264127d90e6810e357b5

**Comment by patcon** [comment]
Thanks vachan! I really appreciate accommodating the feedback :) I'm realizing it feels like maybe a step backward to have the UserTableFilterDialog test checking shapes of the _transformed_ state, as it exists in UserFilterInput (whose state lives up two levels of components in UserTable).

I'll show what I meant with the other way, which keeps outside concerns out of the test (I'll push a branch with the reversion and change, and feel free to yoink those commits into here if it looks ok!)

**Review by brindasasi — CHANGES_REQUESTED** [review]
I don't see the network requests  when I play around with the filters. Fancy filter doesn't filter the rows as it should be. 
Possible it is broken with the latest changes?

**Comment by patcon** [comment]
Hm, will test. Thanks @brindasasi! Just rebuilding now

Thoughts on e2e test for this now that it's working in the app? I'm happy to write it or support writing it, but if we're itching for merge on this, I could spin it out

**Comment by vd1992** [comment]
huh, that is odd, still seems to work on my end

image

image

**Comment by brindasasi** [comment]
hah .. my bad .. clearing cache & everything .. it works now .
In safari there is a spacing issue in there. Chrome looks fine. Is it easy enough to fix in this pr? 

Screen Shot 2022-09-13 at 5 09 19 PM

**Comment by vd1992** [comment]
> hah .. my bad .. clearing cache & everything .. it works now . In safari there is a spacing issue in there. Chrome looks fine. Is it easy enough to fix in this pr?
> 
> Screen Shot 2022-09-13 at 5 09 19 PM

it looks fine in Firefox and in Edge too
hm, I am hesitant about playing around with this because it could be a tricky problem and actually end up being a Hydrogen thing

**Comment by patcon** [comment]
Safari thing was pretty small, so pushed a commit :) thanks B

**Comment by vd1992** [comment]
> Safari thing was pretty small, so pushed a commit :) thanks B

nice!

**Comment by patcon** [comment]
As mentioned, pushed an example of the reversions, and the alt approach:
https://github.com/GCTC-NTGC/gc-digital-talent/pull/3955

(It's not as drastic as it may look, but bringing back the prior tests make it look bigger than it is 😬  Basically just moving the FormValues=>UserFilterInput transformation from UserTableFilterDialog to TableHeader, which is closer to where the UserFilterInput state lives in UserTable component)

When bringing back the very focussed tests from before, I'm not sure how to salvage the tests you'd repurposed from modifying the prior ones. If we want that transformation code tested (and we don't need to), I wonder if it should be tested on its own, rather than trying to wrap it up within the tests of UserTableFilterDialog?

Like everything inside onFilterChange is essentially a transformation that could have it's own function outside the component and be tested separately, and needn't be part of the full component test :)

**Comment by vd1992** [comment]
> As mentioned, pushed an example of the reversions, and the alt approach: #3955
> 
> (It's not as drastic as it may look, but bringing back the prior tests make it look bigger than it is 😬 Basically just moving the FormValues=>UserFilterInput transformation from UserTableFilterDialog to TableHeader, which is closer to where the UserFilterInput state lives in UserTable component)
> 
> When bringing back the very focussed tests from before, I'm not sure how to salvage the tests you'd repurposed from modifying the prior ones. If we want that transformation code tested (and we don't need to), I wonder if it should be tested on its own, rather than trying to wrap it up within the tests of UserTableFilterDialog?
> 
> Like everything inside onFilterChange is essentially a transformation that could have it's own function outside the component and be tested separately, and needn't be part of the full component test :)

I already merged it in lol

**Comment by patcon** [comment]
haha ok, i think it's good to merge if you don't feel like extracting the transformation into a testable `transformFormValuesToUserFilterInput(data)`! I'll just approve and leave it up to you

**Comment by patcon** [comment]
And any feelings on cypress tests here? I could just mark it as `e2e-candidate` if you'd rather get this merged asap?

**Review by patcon — APPROVED** [review]
Approving, as this is blocking and I'd rather see you getting the endorphins of merging if you need it, but I think a cypress test and tests of that transformation could be nice! If you agree, let's track them in other issues (the ones blocked on this) :)

**Comment by vd1992** [comment]
> And any feelings on cypress tests here? I could just mark it as `e2e-candidate` if you'd rather get this merged asap?

I wasn't sure how to meaningfully do a Cypress test here since there are so many filters at play that it is hard to accurately ensure things are working as intended without a thorough manual exercise.

**Comment by vd1992** [comment]
@patcon 
uh-oh, styling broke

image

**Comment by vd1992** [comment]
oops, I'm silly, I shrunk the window that is why the rendering changed 💀

**Comment by patcon** [comment]
> I wasn't sure how to meaningfully do a Cypress test here since there are so many filters at play that it is hard to accurately ensure things are working as intended without a thorough manual exercise.

Happy to defer, but also, this feels like classic place for "don't let perfect be enemy of the good". Even a single check of one field would catch a whole class of wiring-things-together regressions :)

**Review by brindasasi — COMMENTED** [review]
Going into the filter pop up didn't trigger the network calls in my local. 
It works in Pat's local as he mentioned. I'm fine if you go ahead with the granted approval. I can still continue testing on this in Dev environment to make sure.

**Comment by vd1992** [comment]
@patcon 
I added a basic check with https://github.com/GCTC-NTGC/gc-digital-talent/pull/3893/commits/6d47667251a95312a1a00b0c652051ec1c2a6f15
if that is fine by you I will merge this in

**Comment by patcon** [comment]
LGTM! Wooooo

**Pull request merged**
